testcases/kernel/syscalls/recvmmsg/recvmmsg01.c

```diff
@@ -0,0 +1,2389 @@
+							      SOCK_DGRAM);
+
+	call_receive_iovec_boundary_checks_init(&ribc_0, 1);
+	call_receive_iovec_boundary_checks_init(&ribc_1, 2);
+	call_receive_iovec_boundary_checks_init(&ribc_2, 3);
+	call_receive_iovec_boundary_checks_init(&ribc_3, iov_max - 2);
+	call_receive_iovec_boundary_checks_init(&ribc_4, iov_max - 1);
+	call_receive_iovec_boundary_checks_init(&ribc_5, iov_max);
+	call_receive_iovec_boundary_checks_init(&ribc_6, iov_max + 1);
+	call_receive_iovec_boundary_checks_init(&ribc_7, iov_max + 2);
+
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_0, 1);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_1, 2);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_2, 3);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_3, iov_max - 2);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_4, iov_max - 1);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_5, iov_max);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_6, iov_max + 1);
+	call_receive_iovec_boundary_checks_peeking_init(&ribcp_7, iov_max + 2);
+
+	bool some_data = true;
+	bool no_data = false;
+
+	call_receive_file_descriptor_init(&rfd_0, 0, some_data, 0);
+	call_receive_file_descriptor_init(&rfd_1, 0, no_data, 0);
+	call_receive_file_descriptor_init(&rfd_2, MSG_CMSG_CLOEXEC,
+					  some_data, 0);
+	call_receive_file_descriptor_init(&rfd_3, MSG_CMSG_CLOEXEC, no_data, 0);
+
+	/*
+	 * ctlmsgfd_t (with 64 bit size_t) gets an extra int of padding
+	 * missing bytes then has to be sizeof(size_t) not sizeof(int),
+	 * otherwise there is still space for a sizeof(int) sized fd
+	 */
+
+	call_receive_file_descriptor_init(&rfd_4, 0, some_data,
+					  -sizeof(size_t));
+	call_receive_file_descriptor_init(&rfd_5, 0, no_data,
+					  -sizeof(size_t));
+	call_receive_file_descriptor_init(&rfd_6, MSG_CMSG_CLOEXEC, some_data,
+					  -sizeof(size_t));
+	call_receive_file_descriptor_init(&rfd_7, MSG_CMSG_CLOEXEC, no_data,
+					  -sizeof(size_t));
+
+	/* missing actual fields in the underlying cmsghdr_t */
+
+	call_receive_file_descriptor_init(&rfd_8, 0, some_data,
+					  -(ssize_t) (2 * sizeof(size_t)));
+	call_receive_file_descriptor_init(&rfd_9, 0, no_data,
+					  -(ssize_t) (2 * sizeof(size_t)));
+	call_receive_file_descriptor_init(&rfd_10, MSG_CMSG_CLOEXEC, some_data,
+					  -(ssize_t) (2 * sizeof(size_t)));
+	call_receive_file_descriptor_init(&rfd_11, MSG_CMSG_CLOEXEC, no_data,
+					  -(ssize_t) (2 * sizeof(size_t)));
+
+	/* just one byte in the underlying cmsghdr_t */
+
+	call_receive_file_descriptor_init(&rfd_12, 0, some_data,
+					  -(ssize_t) (sizeof(ctlmsgfd_t) - 1));
+	call_receive_file_descriptor_init(&rfd_13, 0, no_data,
+					  -(ssize_t) (sizeof(ctlmsgfd_t) - 1));
+	call_receive_file_descriptor_init(&rfd_14, MSG_CMSG_CLOEXEC, some_data,
+					  -(ssize_t) (sizeof(ctlmsgfd_t) - 1));
+	call_receive_file_descriptor_init(&rfd_15, MSG_CMSG_CLOEXEC, no_data,
+					  -(ssize_t) (sizeof(ctlmsgfd_t) - 1));
+}
```

Comment:
And if it's not possible or easy to track the test variants in statically initialized data, it's probably much easier to just write the actual test functions for each testcase rather than trying to abstract them into several layers of indirections.